File: lib/java_impl/talkhouse/HammingWindow.java
package talkhouse;
import java.util.List;
import java.util.ArrayList;

public class HammingWindow {
    private double[] hammingWindow;
    public HammingWindow(int windowSize) {
        hammingWindow = new double[windowSize]; 
        double twopi = Math.PI * 2;
        for (int i=0;i<windowSize;++i) {
            hammingWindow[i] = 0.54 - 0.46*Math.cos(twopi*i/(windowSize-1));
        }
    }

    public double[][] apply(double[][] data) {
        double[][] result = new double[data.length][];
        for (int i=0;i<data.length;++i) {
            result[i] = new double[data[i].length];
            for (int j=0;j<hammingWindow.length;++j) {
                result[i][j] = hammingWindow[j] * data[i][j];
            }
        }
        return result;
    }
};

File: lib/ruby_impl/hamming_window.rb
module Signal
  class HammingWindow
    include Math
    def initialize window_size
  	  twopi = 2 * PI
  	  @hamming_window = []
      window_size.times do |i|
  	  	@hamming_window << 0.54 - 0.46*cos(twopi*i/(window_size-1))
  	  end
    end
    def << segments
      segments.map do |s|
        s.zip(@hamming_window).map {|d, h| d*h}
      end
    end
  end
end
